lib/OpenQA/Setup.pm

```diff
@@ -328,6 +328,60 @@ sub load_plugins {
     OpenQA::Setup::update_config($server->config, @{$server->plugins->namespaces}, "OpenQA::WebAPI::Auth");
 }
 
+sub load_ui_plugins {
+    my ($server, $plugin_route) = @_;
+
+    return if (!defined $server->config->{ini_config} || !defined $server->config->{ini_config}->{v});
+
+    push @{$server->plugins->namespaces}, 'WebAPIPlugin';
+    for my $section (@{$server->config->{ini_config}->{sects}}) {
+        my $plugin;
+        if ($section =~ /WebAPIPlugin(.+)/) {
+            $plugin = $1;
+        }
+        else {
+            next;
+        }
+        my $plugin_settings = $server->config->{ini_config}->{v}->{$section};
+        next if (!$plugin_settings->{enabled});
+        $server->log->info("Loading plugin $section...");
+        my $res = eval {
+            my $path     = $plugin_settings->{path};
+            my $jsonfile = "$path/WebAPIPlugin/$plugin.json";
+            $server->log->info("Checking $jsonfile...");
+            if (-f $jsonfile) {
+                # load routes if defined
+                $server->plugin(
+                    OpenAPI => {
+                        route => $plugin_route,
+                        url   => $jsonfile
+                    });
+                push @{$server->renderer->paths}, $path . '/WebAPIPlugin';
+                $server->log->info("Loaded $jsonfile");
+            }
+            my $res = eval "use lib '$path';\nuse WebAPIPlugin::$plugin;\n1;";
+            if (!defined $res) {
+                die $@;
+            }
+            # now load actual plugin - should work if command above succeeded
+            $server->plugin($plugin, $plugin_settings);
+            $server->log->info("Loaded plugin $section");
+            1;
+        };
+        my $err = $@;
+        if (!defined $res || $res != 1) {
+            my $msg = "Failed to load $section (enabled=" . $plugin_settings->{enabled} . "): {$err}";
+            $server->log->error($msg);
+            if ($plugin_settings->{enabled} eq "required") {
+                die $msg;
+            }
+            else {
+                warn $msg;
+            }
```

Comment:
This seems redundant with `enabled`.

Reply:
My idea here was that Plugin load warnings are printed back to human when service is started manually.

Reply:
If a plugin is enabled and loading fails the service should just die.

Reply:
I don't agree with this - some plugins must be really optional. It is a shame if downtime happens when some minor plugin fails to load

Reply:
a) Unstable plugins should never make it into production. b) It should be absolutely obvious to whoever updated the system that the update failed. And they should have to disable the plugin in the config to make the service start again.

Reply:
It is not question of quality, because stable software fails to start from time to time as well. And startup may happen without human interaction at all - so it will lead to downtime until somebody reacts to the problem and is lucky to find root cause and spends time to troubleshoot the reason and implement workaround (hopefully correct one - otherwise it may lead to further problems).